### PR TITLE
supporting CRUD for server groups issue #49

### PR DESCRIPTION
**Pull request by danleyden**

### CODE REVIEW

File: lib/fog/openstack/models/compute/server_group.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        attribute :name
+        attribute :policies, :type => :array
+        attribute :members
+
+        VALID_SERVER_GROUP_POLICIES = ['affinity', 'anti-affinity']
+
+        def self.validate_server_group_policy(policy)
+          fail ArgumentError, "#{policy} is an invalid policy... must use one of #{VALID_SERVER_GROUP_POLICIES.join(', ')}" \
+            unless VALID_SERVER_GROUP_POLICIES.include? policy
+          true
```
Comment (plribeiro3000) [inline lib/fog/openstack/models/compute/server_group.rb]: This is confusing imo. You return true if it is valid but raise an exception if its wrong. What do you think of returning true or false only?

File: lib/fog/openstack/requests/compute/create_server.rb [hunk 2]
```diff
@@ -169,6 +169,12 @@ def create_server(name, image_ref, flavor_ref, options = {})
 
           self.data[:last_modified][:servers][server_id] = Time.now
           self.data[:servers][server_id] = mock_data
+
+          if options['os:scheduler_hints'] && options['os:scheduler_hints']['group']
+            grp = self.data[:server_groups][options['os:scheduler_hints']['group']]
```
Comment (plribeiro3000) [inline lib/fog/openstack/requests/compute/create_server.rb]: `grp` stands for group right?
Can you be more explicitamente and use `group` instead?

File: lib/fog/openstack/requests/compute/delete_server.rb [hunk 3]
```diff
@@ -22,6 +22,8 @@ def delete_server(server_id)
               self.data[:last_modified][:servers].delete(server_id)
               self.data[:servers].delete(server_id)
               response.status = 204
+              group_id, _ = data[:server_groups].find { |_id, grp| grp[:members].include?(server_id) }
```
Comment (plribeiro3000) [inline lib/fog/openstack/requests/compute/delete_server.rb]: Could you please take care of this @houndci-bot comment?

File: lib/fog/openstack/models/compute/server_group.rb (hunk 1 above)
Comment (danleyden) [inline lib/fog/openstack/models/compute/server_group.rb]: That could return true / false, but then the places where that is called should raise the ArgumentError unless it was true (less efficient due to having two checks inline and also duplication of that check / raise)

File: lib/fog/openstack/requests/compute/list_server_groups.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+            :path    => 'os-server-groups',
+            :query   => options
+          )
+        end
+      end
+
+      class Mock
+        def list_server_groups(_options = {})
+          groups = data[:server_groups].map do |id, grp|
+            grp.merge('id' => id, 'project_id' => 'test-project', 'user_id' => 'test-user')
```
Comment (plribeiro3000) [inline lib/fog/openstack/requests/compute/list_server_groups.rb]: Use `group` here as well please.

**Comment by plribeiro3000** [comment]
@danleyden looking good besides some small comments.
Thank you for opening this PR!

File: lib/fog/openstack/models/compute/server_group.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+module Fog
+  module Compute
+    class OpenStack
+      class ServerGroup < Fog::OpenStack::Model
+        identity  :id
+        attribute :name
+        attribute :policies, :type => :array
+        attribute :members
+
+        VALID_SERVER_GROUP_POLICIES = ['affinity', 'anti-affinity']
```
Comment (plribeiro3000) [inline lib/fog/openstack/models/compute/server_group.rb]: Could you fix this as well please?

File: lib/fog/openstack/models/compute/server_group.rb (hunk 1 above)
Comment (plribeiro3000) [inline lib/fog/openstack/models/compute/server_group.rb]: Well. fair enough.

**Comment by plribeiro3000** [comment]
@danleyden Could you take a look at the broken build as well please?

**Comment by plribeiro3000** [comment]
@danleyden I can see there is more instances named `grp` but that should not be a blocker. Does this code currently work? Have you tried it?

**Comment by danleyden** [comment]
I have successfully run this code against a deployment of IceHouse. I don't have access to a deployment of anything more recent.

However, from looking at the API documentation, it *should* work with newer versions.

**Comment by plribeiro3000** [comment]
👍

**Comment by plribeiro3000** [comment]
Sadly, i don't have permissions to merge.
Let me try to get it.

**Comment by plribeiro3000** [comment]
Done. Thanks for your contribution @danleyden !

**Comment by ShimiTaNaka** [comment]
Any idea when it will be out?

**Comment by plribeiro3000** [comment]
@ShimiTaNaka I don't have permissions for `fog-openstack` on rubygems.

cc/ @geemus Can you cut a new release or give me access to do so?

**Comment by geemus** [comment]
released 0.1.10

**Pull request merged**